### PR TITLE
Add Auto-saving and Search to Lesson Resource Selection

### PR DESCRIPTION
**Pull request by jonboiser**
### Summary

1. Adds a search bar to Lesson Resource Selection
1. Removes 'Save' button and automatically saves resource changes when selecting resources
1. Adds a select-all button to topics
1. Updates copy on UI and different snackbar notifications
1. Updates some of the copy on the Create Exam Page
1. Fixes whitespace in .feature files

**Search**
search

**Autosave**
autosave

**Create Exam Page**

screen shot 2018-09-17 at 3 02 58 pm

### Reviewer guidance

**Testing search**

1. Search term with results
1. Search term with no results
1. Search from within UI
1. Search using URL (put search term at the very of the URL)
1. Exiting search page should go back to previous page or channels listing

**Testing auto save**

1. Test toggling select-all
1. Toggling single object
1. Resources are kept when going from topic to topic or preview or search
1. Resources are kept when exiting the selection workflow

### References

…

----

### Contributor Checklist

- [x] Contributor has fully tested the PR manually
- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [ x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label

### Reviewer Checklist

- [ ] Automated test coverage is satisfactory
- [ ] Reviewer has fully tested the PR manually
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependencies files were updated (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Link to diff of internal dependency change is included
- [ ] CHANGELOG.rst is updated for high-level changes
- [ ] Contributor is in AUTHORS.rst

### CODE REVIEW

**Comment by rtibbles** [comment]
To quote @navakk9: "BEST NEWS EVER!" - this is great!

I know it may not be possible in the scope of the time available to get this in, but would be good to think about how we can try to reduce code duplication for search once we have also implemented this for exams (and also with the core search functionality too).

The other question is about the filtering and pagination - the filtering is now dependent on the displayed results, which can hide relevant filter options if there are more results than the default page size. There's also no option to paginate. The search results include metadata that can be used here, the core search functionality implements these, for reference.

File: kolibri/plugins/coach/assets/src/views/CoachIndex.vue [hunk 1]
```diff
@@ -21,6 +21,7 @@
       </template>
 
       <component :is="currentPage" />
+      <router-view />
```
Comment (christianmemije) [inline kolibri/plugins/coach/assets/src/views/CoachIndex.vue]: What purpose does this serve here?

File: kolibri/plugins/coach/assets/src/views/assignments/AssignmentSummary.vue [hunk 2]
```diff
@@ -123,7 +123,7 @@
       changeStatus: 'Change',
       description: 'Description',
       noDescription: 'No description',
-      recipients: 'Recipients',
+      assignedGroupsListLabel: 'Visible to',
```
Comment (christianmemije) [inline kolibri/plugins/coach/assets/src/views/assignments/AssignmentSummary.vue:126]: :mood:  But now this does not match the modal...

File: kolibri/plugins/coach/assets/src/views/exams/CreateExamPage/index.vue [hunk 3]
```diff
@@ -3,7 +3,7 @@
   <div>
     <h1>{{ $tr('createNewExam') }}</h1>
     <KGrid>
-      <KGridItem size="4">
+      <KGridItem size="100" percentage>
```
Comment (christianmemije) [inline kolibri/plugins/coach/assets/src/views/exams/CreateExamPage/index.vue]: If these are always going to take up 100% then there is no need to use a grid.

File: kolibri/plugins/coach/assets/src/views/lessons/LessonResourceSelectionPage/SearchTools/LessonsSearchBox.vue [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        return this.searchTerm !== this.$route.params.searchTerm;
+      },
+    },
+    methods: {
+      clearSearchTerm() {
+        this.searchTerm = '';
+        this.$refs.searchinput.focus();
+      },
+      search() {
+        this.$emit('searchterm', this.searchTerm);
```
Comment (christianmemije) [inline kolibri/plugins/coach/assets/src/views/lessons/LessonResourceSelectionPage/SearchTools/LessonsSearchBox.vue]: Should prevent searching for an empty string.

File: kolibri/plugins/coach/assets/src/views/lessons/LessonResourceSelectionPage/index.vue [hunk 5]
```diff
@@ -1,116 +1,251 @@
...
+    <ContentCardList
+      v-if="!isExiting"
+      :contentList="filteredContentList"
+      :showSelectAll="selectAllIsVisible"
+      :selectAllChecked="addableContent.length === 0"
+      :contentIsChecked="contentIsInLesson"
+      :contentHasCheckbox="contentIsDirectoryKind"
+      :contentCardMessage="selectionMetadata"
+      :contentCardLink="contentLink"
+      @change_select_all="toggleTopicInWorkingResources"
```
Comment (christianmemije) [inline kolibri/plugins/coach/assets/src/views/lessons/LessonResourceSelectionPage/index.vue]: :snake:

File: kolibri/plugins/coach/assets/src/views/assignments/AssignmentSummary.vue (hunk 2 above)
Comment (jonboiser) [reply kolibri/plugins/coach/assets/src/views/assignments/AssignmentSummary.vue:126]: We should change the label on the modal then

File: kolibri/plugins/coach/assets/src/views/lessons/LessonResourceSelectionPage/index.vue (hunk 5 above)
Comment (jonboiser) [reply kolibri/plugins/coach/assets/src/views/lessons/LessonResourceSelectionPage/index.vue]: `@changeselectall`?

Comment (christianmemije) [reply kolibri/plugins/coach/assets/src/views/lessons/LessonResourceSelectionPage/index.vue]: `@CHANGE-$ELECTall`

File: integration_testing/features/admin/admin-change-facility-settings.feature [hunk 6]
```diff
@@ -1,10 +1,10 @@
 Feature: Admin change facility settings
-    Admin needs to be able to change the user sign-in/up, self-edit, and content download options according to the needs of the facility 
+  Admin needs to be able to change the user sign-in/up, self-edit, and content download options according to the needs of the facility
```
Comment (indirectlylit) [inline integration_testing/features/admin/admin-change-facility-settings.feature:2]: were all the formatting changes to feature files supposed to be part of this PR?

File: kolibri/plugins/coach/assets/src/views/CoachIndex.vue (hunk 1 above)
Comment (indirectlylit) [inline kolibri/plugins/coach/assets/src/views/CoachIndex.vue:24]: why was this added?

File: kolibri/plugins/coach/assets/src/views/lessons/LessonResourceSelectionPage/SearchTools/LessonsSearchBox.vue [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        :placeholder="$tr('searchBoxLabel')"
+        ref="searchinput"
+      >
+
+      <div class="buttons-wrapper">
+        <UiIconButton
+          color="black"
+          size="small"
+          class="clear-button"
+          :class="searchTerm === '' ? '' : 'clear-button-visble'"
```
Comment (indirectlylit) [inline kolibri/plugins/coach/assets/src/views/lessons/LessonResourceSelectionPage/SearchTools/LessonsSearchBox.vue]: `visble` - typo

File: kolibri/plugins/coach/assets/src/views/lessons/LessonResourceSelectionPage/SearchTools/LessonsSearchFilters.vue [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        type: Array,
+        default() {
+          return [];
+        },
+      },
+      value: {
+        // Map of { kind, channel, role }
+        type: Object,
+        required: true,
+      },
```
Comment (indirectlylit) [inline kolibri/plugins/coach/assets/src/views/lessons/LessonResourceSelectionPage/SearchTools/LessonsSearchFilters.vue:100]: Recommend adding validators on `searchResults` and `value` would be useful since they are not primitive types

**Review by indirectlylit — COMMENTED** [review]
## Notes

This button should say 'Manage resources':

image

'Total resources selected: 2' is confusing with auto-save. Recommend changing to '2 resources in lesson' or similar

image

Similarly here: 

image

Instead of 'Undo', recommend this link says 'Remove' because it could have happened a long time ago

image

File: integration_testing/features/admin/admin-change-facility-settings.feature (hunk 6 above)
Comment (indirectlylit) [reply integration_testing/features/admin/admin-change-facility-settings.feature:2]: whoops I see your note above

File: kolibri/plugins/coach/assets/src/modules/lessonResources/handlers.js [hunk 9]
```diff
@@ -226,3 +200,20 @@ function _prepLessonContentPreview(store, classId, lessonId, contentId) {
...
   );
 }
+
+export function showLessonResourceSearchPage(store, params) {
+  return store.dispatch('loading').then(() => {
+    return ContentNodeSearchResource.fetchCollection({
+      getParams: {
+        search: params.searchTerm,
+      },
+    }).then(results => {
```
Comment (rtibbles) [inline kolibri/plugins/coach/assets/src/modules/lessonResources/handlers.js:210]: To do filtering that relies on the entire search set, just need to destructure assignment here:
`{ results, channel_ids, content_kinds, total_results }` instead of `results`.

File: kolibri/plugins/coach/assets/src/modules/lessonResources/handlers.js [hunk 10]
```diff
@@ -226,3 +200,20 @@ function _prepLessonContentPreview(store, classId, lessonId, contentId) {
...
+      return showResourceSelectionPage(store, {
+        classId: params.classId,
```
Comment (rtibbles) [inline kolibri/plugins/coach/assets/src/modules/lessonResources/handlers.js:212]: Then would just need to pass `channel_ids`, `content_kinds` and `total_results` to the state and use it to populate the filters. Could potentially add an additional return property from the search for the coach content filter also.

**Review by indirectlylit — DISMISSED** [review]
seems to work well. pushed some updates addressing feedback so we can get it merged in for string freeze

File: kolibri/plugins/coach/assets/src/views/CoachIndex.vue (hunk 1 above)
Comment (indirectlylit) [reply kolibri/plugins/coach/assets/src/views/CoachIndex.vue:24]: from @jonboiser:

it was necessary to implement `beforeRouteLeave` in-component guards

**Review by indirectlylit — APPROVED** [review]
creating follow-up issue #4297 - @rtibbles feel free to add notes there
